### PR TITLE
Execute crowbar upgrade prechecks

### PR DESCRIPTION
**Pull request by jsuchome**
Grepping for result looks ugly, I think it would be better if crowbarcrl exits with non-0 in the failed check cases. Or is there any better option?

### CODE REVIEW

**Comment by jsuchome** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;BfMFXB3r&#x2F;64-5-s47p5-s48p19-add-ci-jobs-for-new-upgrade-process-with-and-without-ha"> (5) S47P5 (S48P19) Add CI jobs for new upgrade process with and without HA</a>

**Comment by jsuchome** [comment]
@MaximilianMeister ok, added; also it still requires grepping that output.

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -4586,6 +4586,14 @@ function onadmin_prepare_crowbar_upgrade()
         # move nodes to upgrade mode
         safely curl -s -X POST $crowbar_api_digest $crowbar_api/installer/upgrade/prepare.json
     else
+        if crowbarctl upgrade prechecks --format plain | grep "false" ; then
+            complain 11 "Some necessary check before the upgrade has failed"
+        fi
```
Comment (MaximilianMeister) [inline scripts/qa_crowbarsetup.sh]: another way would be to use our internal json parser:

```bash
crowbarctl upgrade prechecks --format json | rubyjsonparse 'puts "failed" if j.keys.select { |val| j[val] == false }.any?'
```

... and then check for "failed" as a return

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Is it really more readable? Wouldn't it be better if crowbarctl would exit non-zero value?

Comment (MaximilianMeister) [inline scripts/qa_crowbarsetup.sh]: you could even find out that way which check failed (with another ruby command of course)

```bash
crowbarctl upgrade prechecks --format json | rubyjsonparse 'j.keys.each { |val| puts "#{val} failed" unless j[val] }'
```

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: That grep shows that information in mkcloud log as well. So I'd rather go for simple solution.

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -4586,6 +4586,14 @@ function onadmin_prepare_crowbar_upgrade()
...
+
+        # FIXME crowbatctl is currently missing admin repo check
```
Comment (MaximilianMeister) [inline scripts/qa_crowbarsetup.sh]: we just released a new rubygem for `crowbar-client`, so the subcommand should be in the product today.

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Thanks, I'll update the PR once I can test it

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -4586,6 +4586,14 @@ function onadmin_prepare_crowbar_upgrade()
...
+        if curl -s -X GET -H "Accept: application/vnd.crowbar.v2.0+json" $crowbar_api/api/crowbar/repocheck | grep "missing" ; then
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: Is it possible to use the new `crowbar_api_request` function here?
If the http code is non [23]xx then the function fails and could thus be used in the if clause.
So the api endpoint needs to return a proper http code - does it do that, @MaximilianMeister ?

Comment (MaximilianMeister) [inline scripts/qa_crowbarsetup.sh]: we won't need it anyways, we can use `crowbarctl`

**Comment by jsuchome** [comment]
Updated to include repo checks via crowbarctl (and tested)

File: scripts/qa_crowbarsetup.sh (hunk 1 above)
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: I agree it's not nice to require parsing in order to determine a binary result (success or failure). That's exactly what exit codes are supposed to be for, after all.

**Pull request merged**